protocol/tls.go

```diff
@@ -0,0 +1,134 @@
+			if err != nil {
+				return nil, fmt.Errorf("(%s) error combining fetch node creds next proto value", op)
+			}
+			// Get the raw proto bytes
+			reqBytes, err := base64.RawStdEncoding.DecodeString(protoString)
+			if err != nil {
+				return nil, fmt.Errorf("(%s) error base64-decoding fetch node creds next proto value", op)
+			}
+			// Decode the proto into the request
+			req := new(types.FetchNodeCredentialsRequest)
+			if err := proto.Unmarshal(reqBytes, req); err != nil {
+				return nil, fmt.Errorf("(%s) error unmarshaling common name value: %w", op, err)
+			}
+			// This will return a response either with Authorized false and no
+			// other data or Authorized true and encrypted values
+			fetchResp, err := l.fetchCredsFn(l.ctx, l.storage, req, l.options...)
+			if err != nil {
+				return nil, fmt.Errorf("(%s) error handling fetch creds: %w", op, err)
+			}
+			fetchRespBytes, err := proto.Marshal(fetchResp)
+			if err != nil {
+				return nil, fmt.Errorf("(%s) error marshaling fetch response: %w", op, err)
+			}
+			// Have the response put into the common name
+			serverCertsReq.CommonName = base64.RawStdEncoding.EncodeToString(fetchRespBytes)
+
+			// This is a bit redundant with the fetch function above but we need
+			// a few values
+			var reqInfo types.FetchNodeCredentialsInfo
+			if err := proto.Unmarshal(req.Bundle, &reqInfo); err != nil {
+				return nil, fmt.Errorf("(%s) cannot unmarshal request info: %w", op, err)
+			}
+
+			// We are returning either unauthorized or encrypted creds so we
+			// don't want to validate the node's credentials -- it doesn't have
+			// them yet!
+			serverCertsReq.SkipVerification = true
+			serverCertsReq.Nonce = reqInfo.Nonce
+			serverCertsReq.CertificatePublicKeyPkix = reqInfo.CertificatePublicKeyPkix
+			protoToReturn = p
+
+		case strings.HasPrefix(p, nodeenrollment.AuthenticateNodeNextProtoV1Prefix):
+			// Get the full string and pull out just the marshaled proto
+			protoString, err := nodetls.CombineFromNextProtos(nodeenrollment.AuthenticateNodeNextProtoV1Prefix, hello.SupportedProtos)
+			if err != nil {
+				return nil, fmt.Errorf("(%s) error combining auth node next proto value", op)
+			}
+			// Get the raw proto bytes
+			reqBytes, err := base64.RawStdEncoding.DecodeString(protoString)
+			if err != nil {
+				return nil, fmt.Errorf("(%s) error base64-decoding auth node next proto value", op)
+			}
+			// Decode the certs request
+			if err := proto.Unmarshal(reqBytes, serverCertsReq); err != nil {
+				return nil, fmt.Errorf("(%s) error unmarshaling common name value: %w", op, err)
+			}
+			protoToReturn = p
+
+		default:
+			continue
+		}
+
+		// If we're here we found one of our known protos so break out of
+		// the for loop and finish
+		break
+	}
+
+	// Generate a server-side certificate
+	certResp, err := l.generateServerCertificatesFn(l.ctx, l.storage, serverCertsReq, l.options...)
```

Comment:
I might be tempted to add a check for certResp != nil since it's a caller supplied func implementation.

Reply:
Good idea, fixed